src/Infrastructure/Multitenancy/TenantService.cs

```diff
@@ -0,0 +1,111 @@
+    {
+        _tenantStore = tenantStore;
+        _csSecurer = csSecurer;
+        _dbInitializer = dbInitializer;
+        _localizer = localizer;
+    }
+
+    public async Task<List<TenantDto>> GetAllAsync()
+    {
+        var tenants = (await _tenantStore.GetAllAsync()).Adapt<List<TenantDto>>();
+
+        tenants.ForEach(t => t.ConnectionString = _csSecurer.MakeSecure(t.ConnectionString));
+
+        return tenants;
+    }
+
+    public async Task<bool> ExistsWithIdAsync(string id) =>
+        await _tenantStore.TryGetAsync(id) is not null;
+
+    public async Task<bool> ExistsWithNameAsync(string name) =>
+        (await _tenantStore.GetAllAsync()).Any(t => t.Name == name);
+
+    public async Task<TenantDto> GetByIdAsync(string id) =>
+        (await GetTenantInfoAsync(id))
+            .Adapt<TenantDto>();
+
+    public async Task<string> CreateAsync(CreateTenantRequest request, CancellationToken cancellationToken)
+    {
+        var tenant = new FSHTenantInfo(request.TenantId, request.Name, request.ConnectionString, request.AdminEmail, request.Identifier);
+
+        await _tenantStore.TryAddAsync(tenant);
+
+        try
+        {
+            // TODO: we should probably run this in a hangfire job?
+            await _dbInitializer.InitializeApplicationDbForTenantAsync(tenant, cancellationToken);
+        }
+        catch
+        {
+            await _tenantStore.TryRemoveAsync(request.TenantId);
+            throw;
+        }
+
+        return tenant.Id;
+    }
+
+    public async Task<string> ActivateAsync(string id)
+    {
+        var tenant = await GetTenantInfoAsync(id);
+
+        if (tenant.IsActive)
+        {
+            throw new ConflictException("Tenant is already Activated.");
+        }
+
+        tenant.Activate();
+
+        await _tenantStore.TryUpdateAsync(tenant);
+
+        return $"Tenant {id} is now Activated.";
+    }
+
+    public async Task<string> DeactivateAsync(string id)
+    {
+        var tenant = await GetTenantInfoAsync(id);
+
+        if (!tenant.IsActive)
+        {
+            throw new ConflictException("Tenant is already Deactivated.");
+        }
+
+        tenant.Deactivate();
+
+        await _tenantStore.TryUpdateAsync(tenant);
+
+        return $"Tenant {id} is now Deactivated.";
+    }
+
+    public async Task<string> UpdateSubscription(string id, DateTime extendedExpiryDate)
+    {
+        var tenant = await GetTenantInfoAsync(id);
+
+        tenant.SetValidity(extendedExpiryDate);
+
+        await _tenantStore.TryUpdateAsync(tenant);
+
+        return $"Tenant {id}'s Subscription Upgraded. Now Valid till {tenant.ValidUpto}.";
+    }
+
+    private async Task<FSHTenantInfo> GetTenantInfoAsync(string id) =>
+        await _tenantStore.TryGetAsync(id)
+            ?? throw new NotFoundException(string.Format(_localizer["entity.notfound"], typeof(FSHTenantInfo).Name, id));
```

Comment:
tenant*.notfound

Reply:
that's what it says actually... it's using the generic entity.notfound which has 2 parameters, the name of the entitytype and the id... but yeah would probably be better with a specific one...

The localizer strings is also something that could use some "re-organising"  in general ;-)
and more consistency...

Reply:
I ran a search and there's no "tenant.notfound" used yet anywere.
Let's keep this for another pr... maybe should create an issue for it for another PR... I created an issue https://github.com/fullstackhero/dotnet-webapi-boilerplate/issues/372

Reply:
I might start to sound like a broken record, but I would personally go for the standard .NET way of working with resources. Which has a lot of tooling and support built around it...
Like you also do on the blazor project...